libr/anal/type.c

```diff
@@ -120,20 +120,258 @@ R_API void r_anal_save_parsed_type(RAnal *anal, const char *parsed) {
+	free (key);
+
+	RVector members;
+	r_vector_init (&members, sizeof (RAnalStructMember), struct_type_fini, NULL);
+
+	if (!r_vector_reserve (&members, (size_t)sdb_alen (sdb_members))) {
+		goto error;
+	}
+
+	char *cur;
+	sdb_aforeach (cur, sdb_members) {
+		char *type_key = r_str_newf ("struct.%s.%s", sanitized_name, cur);
+		if (!type_key) {
+			goto error;
+		}
+		char *values = sdb_get (anal->sdb_types, type_key, NULL);
+		free (type_key);
+
+		if (!values) {
+			goto error;
+		}
+		char *offset = NULL;
+		char *type = sdb_anext (values, &offset);
+		if (!offset) { // offset is missing, malformed state
+			free (values);
+			goto error;
+		}
+		offset = sdb_anext (offset, NULL);
+		RAnalStructMember cas = {
+			 .name = strdup (cur), 
+			 .type = strdup (type), 
+			 .offset = strtol (offset, NULL, 10) };
+
+		free (values);
+
+		void *element = r_vector_push (&members, &cas); // returns null if no space available
+		if (!element) {
+			goto error;
+		}
+
+		sdb_aforeach_next (cur);
+	}
+	free (sdb_members);
+	base_struct.members = members;
+	base_type->struct_data = base_struct;
+
+	return base_type;
+
+error:
+	free (base_type);
+	free (sdb_members);
+	r_vector_fini (&members);
+	return NULL;
+}
+
+static RAnalBaseType *get_union_type(RAnal *anal, const char *sanitized_name) {
+	r_return_val_if_fail (anal && sanitized_name, NULL);
+
+	RAnalBaseType *base_type = R_NEW0 (RAnalBaseType);
+	if (!base_type) {
+		return NULL;
+	}
+	base_type->kind = R_ANAL_BASE_TYPE_KIND_UNION;
+
+	RAnalBaseTypeUnion base_union;
+	char *key = r_str_newf ("union.%s", sanitized_name);
+	if (!key) {
+		free (base_type);
+		return NULL;
+	}
+	char *sdb_members = sdb_get (anal->sdb_types, key, NULL);
+	free (key);
+
+	RVector members;
+	r_vector_init (&members, sizeof (RAnalUnionMember), union_type_fini, NULL);
+
+	if (!r_vector_reserve (&members, (size_t)sdb_alen (sdb_members))) {
+		goto error;
+	}
+
+	char *cur;
+	sdb_aforeach (cur, sdb_members) {
+		char *type_key = r_str_newf ("union.%s.%s", sanitized_name, cur);
+		if (!type_key) {
+			goto error;
+		}
+		char *values = sdb_get (anal->sdb_types, type_key, NULL);
+		free (type_key);
+
+		if (!values) {
+			goto error;
+		}
+		char *value = sdb_anext (values, NULL);
+		RAnalUnionMember cas = { .name = strdup (cur), .type = strdup (value) };
+		free (values);
+
+		void *element = r_vector_push (&members, &cas); // returns null if no space available
+		if (!element) {
+			goto error;
+		}
+
+		sdb_aforeach_next (cur);
+	}
+	free (sdb_members);
+	base_union.members = members;
+	base_type->union_data = base_union;
+
+	return base_type;
+
+error:
+	free (base_type);
+	free (sdb_members);
+	r_vector_fini (&members);
+	return NULL;
+}
+
+// returns NULL if name is not found or any failure happened
+R_API RAnalBaseType *r_anal_get_base_type(RAnal *anal, const char *name) {
+	r_return_val_if_fail (anal && name, NULL);
+
+	char *name_sanitized = r_str_sanitize_sdb_key (name);
```

Comment:
lexically should be sanitized_name, but you can use  `sname` which is shorter and looks readable to me

Reply:
Sure, I've went with `sname` then